<a href="https://colab.research.google.com/github/IA1819/entregable2-lucabem/blob/master/AnalisisDataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import numpy as np
import pandas as pd

In [48]:
from google.colab import files
uploaded = files.upload()

Saving sample.csv to sample (1).csv
Saving test.csv to test (1).csv
Saving train.csv to train (1).csv


In [0]:
import io
df_sample = pd.read_csv(io.BytesIO(uploaded['sample.csv']))
df_train = pd.read_csv(io.BytesIO(uploaded['train.csv']))
df_test = pd.read_csv(io.BytesIO(uploaded['test.csv']))

In [50]:
df_sample.head()

Id  Prediction
0   0        True
1   1        True
2   2        True
3   3        True
4   4        True

In [51]:
df_train.head()

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
0   40    Male              1.9               1.0                   231   
1   64    Male              0.9               0.3                   310   
2   40  Female              2.1               1.0                   768   
3   32    Male             25.0              13.7                   560   
4   48    Male              0.7               0.2                   326   

   Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
0                        16                          55             4.3   
1                        61                          58             7.0   
2                        74                         141             7.8   
3                        41                          88             7.9   
4                        29                          17             8.7   

   Albumin  Albumin_and_Globulin_Ratio  Prediction  
0      1.6                         0.6       False  
1      3.4                         0.9        True  
2      4.9                         1.6       False  
3      2.5                         2.5       False  
4      5.5                         1.7       False

In [52]:
df_test.head()

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
0   36    Male              0.8               0.2                   158   
1   55    Male             14.1               7.6                   750   
2   61    Male              0.8               0.2                   192   
3   46    Male              9.4               5.2                   268   
4   42  Female              2.3               1.1                   292   

   Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
0                        29                          39             6.0   
1                        35                          63             5.0   
2                        28                          35             6.9   
3                        21                          63             6.4   
4                        29                          39             4.1   

   Albumin  Albumin_and_Globulin_Ratio  
0      2.2                        0.50  
1      1.6                        0.47  
2      3.4                        0.90  
3      2.8                        0.80  
4      1.8                        0.70

# Análisis del Datasets

La siguiente función nos muestra la dimensión del dataset.

In [53]:
df_train.shape

(437, 11)

El conjunto de datos para entrenar nuestro modelo tiene 437 instancia, cada una evaluada por 11 descriptores. Vamos a ver los distintos descriptores que tiene nuestro conjunto de datos.

La siguiente función nos muestra los descriptores de nuestro dataset.

In [54]:
df_train.columns

Index(['Age', 'Gender', 'Total_Bilirubin', 'Direct_Bilirubin',
       'Alkaline_Phosphotase', 'Alamine_Aminotransferase',
       'Aspartate_Aminotransferase', 'Total_Protiens', 'Albumin',
       'Albumin_and_Globulin_Ratio', 'Prediction'],
      dtype='object')

También podemos ver el tipo de cada uno de los descriptores, algo que nos servirá más adelante para , por ejemplo, decidir qué descriptores son más importantes.

In [55]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 437 entries, 0 to 436
Data columns (total 11 columns):
Age                           437 non-null int64
Gender                        437 non-null object
Total_Bilirubin               437 non-null float64
Direct_Bilirubin              437 non-null float64
Alkaline_Phosphotase          437 non-null int64
Alamine_Aminotransferase      437 non-null int64
Aspartate_Aminotransferase    437 non-null int64
Total_Protiens                437 non-null float64
Albumin                       437 non-null float64
Albumin_and_Globulin_Ratio    437 non-null float64
Prediction                    437 non-null bool
dtypes: bool(1), float64(5), int64(4), object(1)
memory usage: 34.6+ KB


A parte de los tipos, podemos aceder a datos estadisticos, propios de cada columna

In [56]:
df_train.describe()

Age  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
count  437.000000       437.000000        437.000000            437.000000   
mean    44.881007         3.291762          1.481465            290.757437   
std     16.382597         6.377841          2.794080            243.309400   
min      4.000000         0.400000          0.100000             63.000000   
25%     33.000000         0.800000          0.200000            175.000000   
50%     45.000000         1.000000          0.300000            208.000000   
75%     58.000000         2.600000          1.300000            298.000000   
max     90.000000        75.000000         19.700000           2110.000000   

       Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
count                437.000000                  437.000000      437.000000   
mean                  84.057208                  114.210526        6.524714   
std                  200.949407                  322.074848        1.069224   
min                   10.000000                   10.000000        2.800000   
25%                   23.000000                   25.000000        5.800000   
50%                   35.000000                   42.000000        6.600000   
75%                   60.000000                   85.000000        7.200000   
max                 2000.000000                 4929.000000        9.600000   

          Albumin  Albumin_and_Globulin_Ratio  
count  437.000000                  437.000000  
mean     3.180092                    0.958238  
std      0.778667                    0.316748  
min      1.400000                    0.300000  
25%      2.600000                    0.750000  
50%      3.100000                    1.000000  
75%      3.800000                    1.100000  
max      5.500000                    2.800000

Para ver estadísticas de descriptores no númericos, es necesario indicar explícitamente los tipos de datos que nos interesan en el parámetro include.

In [57]:
df_train.describe(include=['object', 'bool'])

Gender Prediction
count     437        437
unique      2          2
top      Male      False
freq      327        310

Para cambiar el tipo de una columna podemos hacer uso de la siguiente función.

In [0]:
df_train['nombreColumna'] = df_train['nombreColumna'].astype('tipoColumna')

### **Ordenando**

Un DataFrame se puede ordenar por el valor de uno de sus descriptores. Por ejemplo, podemos ordenar nuestro dataset por el valor de Total day charge (usamos ascending=False para ordenar en orden decreciente):

In [59]:
df_train.sort_values(by='Age', ascending=False).head(7)

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
158   90    Male              1.1               0.3                   215   
237   85  Female              1.0               0.3                   208   
90    84  Female              0.7               0.2                   188   
63    78    Male              1.0               0.3                   152   
263   75    Male              1.4               0.4                   215   
172   75    Male              2.9               1.3                   218   
79    75    Male              8.0               4.6                   386   

     Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
158                        46                         134             6.9   
237                        17                          15             7.0   
90                         13                          21             6.0   
63                         28                          70             6.3   
263                        50                          30             5.9   
172                        33                          37             3.0   
79                         30                          25             5.5   

     Albumin  Albumin_and_Globulin_Ratio  Prediction  
158      3.0                        0.70       False  
237      3.6                        1.00        True  
90       3.2                        1.10        True  
63       3.1                        0.90       False  
263      2.6                        0.70       False  
172      1.5                        1.00       False  
79       1.8                        0.48       False


También es posible ordenar por múltiples descriptores.

In [60]:
df_train.sort_values(by=['Total_Bilirubin', 'Direct_Bilirubin'],
        ascending=[True, False]).head(7)

Age  Gender  Total_Bilirubin  Direct_Bilirubin  Alkaline_Phosphotase  \
137   58    Male              0.4               0.1                   100   
51     7    Male              0.5               0.1                   352   
124   17  Female              0.5               0.1                   206   
200   63    Male              0.5               0.1                   170   
222   42  Female              0.5               0.1                   162   
370   60    Male              0.5               0.1                   500   
5     27    Male              0.6               0.2                   161   

     Alamine_Aminotransferase  Aspartate_Aminotransferase  Total_Protiens  \
137                        59                         126             4.3   
51                         28                          51             7.9   
124                        28                          21             7.1   
200                        21                          28             5.5   
222                       155                         108             8.1   
370                        20                          34             5.9   
5                          27                          28             3.7   

     Albumin  Albumin_and_Globulin_Ratio  Prediction  
137      2.5                        1.40       False  
51       4.2                        1.10        True  
124      4.5                        1.70        True  
200      2.5                        0.80       False  
222      4.0                        0.90       False  
370      1.6                        0.37        True  
5        1.6                        0.76        True

### Preguntando al Dataset

Un DataFrame se puede indexar de diferentes maneras.

Para obtener una única fila, se puede usar la construcción DataFrame['Name']. Vamos a usar esta construcción para responder a la pregunta de **¿Cuál es la edad media de nuestro dataset?**

In [61]:
df_train['Age'].mean()

44.88100686498856

La edad media de nuestro conjunto es de casi 45 años. 

Además podemos consultar datos en función de un condicional. La sintaxix es **df[P(df['Name'])]**, donde  P es alguna condición lógica que es comprobada para cada elemento de la columna Name. El resultado de dicho indexado es el Dataframe que consta solo de las filas que satisfacen la condición P en la columna Name.

Por ejemplo, **¿Cual es media de edad de las mujeres?**

In [62]:
datosMujeres = df_train[df_train['Gender'] == 'Female']
datosMujeres['Age'].mean()

42.20909090909091

**¿Cual es la media de bilirrubina total que tienen los hombres? Da también la desviacion tipica**

In [63]:
datosMujeres['Total_Bilirubin'].mean()

2.351818181818181

In [64]:
datosMujeres['Total_Bilirubin'].std()

4.9373084785650585

**¿Cual es la media de los atributos de los hombres que padecen la enfermedad?**

In [0]:
datosHombres = df_train[df_train['Gender'] == 'Male']
hombresEnfermos = datosHombres[datosHombres['Prediction'] == True].mean()

In [66]:
hombresEnfermos

Age                            41.364706
Total_Bilirubin                 1.287059
Direct_Bilirubin                0.482353
Alkaline_Phosphotase          239.823529
Alamine_Aminotransferase       36.341176
Aspartate_Aminotransferase     48.282353
Total_Protiens                  6.587059
Albumin                         3.337647
Albumin_and_Globulin_Ratio      1.016706
Prediction                      1.000000
dtype: float64

Ahora, trata de hacerlo en una linea

In [67]:
df_train[(df_train['Gender'] == 'Male') & (df_train['Prediction'] == True)].mean()

Age                            41.364706
Total_Bilirubin                 1.287059
Direct_Bilirubin                0.482353
Alkaline_Phosphotase          239.823529
Alamine_Aminotransferase       36.341176
Aspartate_Aminotransferase     48.282353
Total_Protiens                  6.587059
Albumin                         3.337647
Albumin_and_Globulin_Ratio      1.016706
Prediction                      1.000000
dtype: float64

**¿Edad de la paciente mas joven que esté enferma?**

In [68]:
df_train[(df_train['Gender'] == 'Female') & (df_train['Prediction'] == True)]['Age'].min()

17


### Aplicando funciones a las celdas, columnas y filas


Para aplicar funciones a una columna se usa el método apply(). Por ejemplo, a continuación mostramos cómo obtener el valor máximo de los distintos descriptores del dataset.

In [69]:
df_train.apply(np.max)

Age                             90
Gender                        Male
Total_Bilirubin                 75
Direct_Bilirubin              19.7
Alkaline_Phosphotase          2110
Alamine_Aminotransferase      2000
Aspartate_Aminotransferase    4929
Total_Protiens                 9.6
Albumin                        5.5
Albumin_and_Globulin_Ratio     2.8
Prediction                    True
dtype: object

**¿Cual es la relaccion entre estar enfermo y el sexo?**

In [70]:
pd.crosstab(df_train['Gender'], df_train['Prediction'])

Prediction  False  True 
Gender                  
Female         68     42
Male          242     85